gluten-core/src/main/scala/org/apache/spark/util/Iterators.scala

```diff
@@ -0,0 +1,154 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.util
+
+import org.apache.spark.{InterruptibleIterator, TaskContext}
+import org.apache.spark.sql.execution.metric.SQLMetric
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
+
+private class PayloadCloser[A](in: Iterator[A])(closeCallback: A => Unit) extends Iterator[A] {
+  private var closer: Option[() => Unit] = None
+
+  TaskResources.addRecycler("Iterators#PayloadCloser", 100) {
+    tryClose()
+  }
+
+  override def hasNext: Boolean = {
+    tryClose()
+    in.hasNext
+  }
+
+  override def next(): A = {
+    val a: A = in.next()
+    closer.synchronized {
+      closer = Some(
+        () => {
+          closeCallback.apply(a)
+        })
+    }
+    a
+  }
+
+  private def tryClose(): Unit = {
+    closer.synchronized {
+      closer match {
+        case Some(c) => c.apply()
+        case None =>
+      }
+      closer = None // make sure the payload is closed once
+    }
+  }
+}
+
+private class IteratorCompleter[A](in: Iterator[A])(completionCallback: => Unit)
+  extends Iterator[A] {
+  private val completed = new AtomicBoolean(false)
+
+  TaskResources.addRecycler("Iterators#IteratorRecycler", 100) {
+    tryComplete()
+  }
+
+  override def hasNext: Boolean = {
+    val out = in.hasNext
+    if (!out) {
+      tryComplete()
+    }
+    out
+  }
+
+  override def next(): A = {
+    in.next()
+  }
+
+  private def tryComplete(): Unit = {
+    if (!completed.compareAndSet(false, true)) {
+      return // make sure the iterator is completed once
+    }
+    completionCallback
+  }
+}
+
+private class PipelineTimeAccumulator[A](in: Iterator[A], pipelineTime: SQLMetric)
+  extends Iterator[A] {
+  private val accumulatedTime: AtomicLong = new AtomicLong(0L)
+
+  TaskResources.addRecycler("Iterators#PipelineTimeAccumulator", 100) {
+    tryFinish()
+  }
+
+  override def hasNext: Boolean = {
+    val prev = System.nanoTime()
+    val out = in.hasNext
+    accumulatedTime.addAndGet(System.nanoTime() - prev)
+    if (!out) {
+      tryFinish()
+    }
+    out
+  }
+
+  override def next(): A = {
+    val prev = System.nanoTime()
+    val out = in.next()
+    accumulatedTime.addAndGet(System.nanoTime() - prev)
+    out
+  }
+
+  private def tryFinish(): Unit = {
+    pipelineTime += accumulatedTime.getAndSet(
+      0L
+    ) // make sure the accumulated time is submitted once
+  }
+}
+
+class WrapperBuilder[A](in: Iterator[A]) { // FIXME how to make the ctor companion-private?
+  private var wrapped: Iterator[A] = in
+
+  def recyclePayload(closeCallback: (A) => Unit): WrapperBuilder[A] = {
```

Comment:
thank you @zhztheplayer for the refactor. Can you explain a bit more why naming a new word payload instead of ColumnarBatch in java ? Do you mean there is something else besides ColumnarBatch will call this method ? I'm wondering if we can make a `recycleColumnarBatch` method to make it clear.

Reply:
In `Iterator[A]`, `A` should be type of "payloads" of the iterator. It's also possible that `A` is not `ColumnarBatch`, a typical case is C2R that creates a row iterator as output. 

Does `payload` sound very confusing here? `element` could be another choice but I felt it probably sounds too general for Gluten's use case.